spring-cloud-dataflow-admin/src/main/java/org/springframework/cloud/dataflow/admin/controller/StreamController.java

```diff
@@ -252,6 +251,37 @@ private void deployStream(StreamDefinition stream, Map<String, String> cumulated
 		}
 	}
 
+	private ModuleType determineModuleType(ModuleDefinition moduleDefinition, int index, boolean hasNext) {
```

Comment:
We are trying to avoid "type" being part of a `ModuleDefinition` since it's really only relevant at the time of deployment, and because we will want to support more flexible topologies (than linear pipelines) in the future and keeping ModuleDefinition simpler will likely facilitate that. On the code-completion side, I would expect that you could rely on the info in the `ModuleRegistration` instead?

Regarding the name of the `hasNext` parameter, I agree we could rename to something clearer. I considered a few names and then went along with the `hasNext` since it matches the method on the `Iterator` where the boolean originates. That said, as we cross a method boundary, it would be better to clarify the ordering. Perhaps we could change it to `hasMoreUpstream` or something. Suggestions?